### PR TITLE
refactor(hummock manager): make `sync group` be part of transaction

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

When we create new compaction groups or drop existing compaction groups, we will create a transaction to alter groups and release a new `HummockVersion` to indicate this change simultaneously.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/meta/src/hummock/manager/mod.rs [hunk 1]
```diff
@@ -1380,36 +1404,38 @@ where
...
-        let (mut new_version_delta, mut new_hummock_version) =
-            match self.sync_group(versioning, &compaction_groups).await? {
-                Some((entry_k, entry_v, new_hummock_version)) => (
-                    BTreeMapEntryTransaction::new_insert(
-                        &mut versioning.hummock_version_deltas,
-                        entry_k,
-                        entry_v,
-                    ),
-                    new_hummock_version,
+        let (mut new_version_delta, mut new_hummock_version) = match self
```
Comment (zwang28) [inline src/meta/src/hummock/manager/mod.rs:1407]: Why do we still need sync_group in commit_epoch, given that we've done that in register/unregister in CompactionGroupManager?

Comment (soundOfDestiny) [reply src/meta/src/hummock/manager/mod.rs:1407]: > Why do we still need sync_group in commit_epoch, given that we've done that in register/unregister in CompactionGroupManager?

We can remove `sync_group` here.

**Comment by hzxa21** [comment]
This PR makes group creation in compaction group manager and hummock version atomic, which is a great improvement. I am thinking about whether we can further simplify group creation: can we avoid reverting a compaction group creation completely and thus avoid leaking hummock manager in stream manager?

**Comment by hzxa21** [comment]
IIUC, currently a new compaction group can only be created during table/mv creation (we can support group split via RPC in the future but just ignore that for now), which means compaction group creation always comes with a barrier. In this case, if we can make the following three operations atomic, group creation will never be reverted:
1. `CompactionGroupManager::register_table_ids`
2. `HummockManager::sync_group`
3. `HummockManager::commit_epoch`

Given that CN can write table state without knowing the exact compaction group id, making the above operations atomic becomes possible. Here is my idea:
- Put a TableId to (CompcationGroupId, TableOption) map as part of the `CreateStreamingJob` [command](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/command.rs#L72)
- After [completing MV creation barrier collection](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/mod.rs#L829), pass the above map into `commit_epoch`.
- Inside `commit_epoch`, do the following operations in sequence:
1. init a etcd transaction
2. register table ids in compaction group manager and put it into the transaction.
3. create GroupConstruct version delta, apply to hummock version and put it into the transaction.
4. execute existing logics in commit_epoch
5. commit the ectd transaction

With this, we can safely avoid reverting group creation during MV creation and avoid storing a ref to HummockManager in stream manager.

File: src/meta/src/hummock/manager/mod.rs [hunk 2]
```diff
@@ -115,10 +115,10 @@ pub type HummockManagerRef<S> = Arc<HummockManager<S>>;
 /// by the way
 /// After called, the `ValTransaction` will be dropped.
 macro_rules! commit_multi_var {
-    ($hummock_mgr:expr, $context_id:expr, $($val_txn:expr),*) => {
+    ($hummock_mgr:expr, $context_id:expr, $trx_extern_part:expr, $($val_txn:expr),*) => {
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/mod.rs:118]: After adding the extra expr in the macro, we need to pass None when calling `commit_multi_var!` in many places. This hurts readibility. Can we pass `Transaction::default()` explicitly instead if we don't need to reuse a txn?

Comment (soundOfDestiny) [reply src/meta/src/hummock/manager/mod.rs:118]: > After adding the extra expr in the macro, we need to pass None when calling `commit_multi_var!` in many places. This hurts readibility. Can we pass `Transaction::default()` explicitly instead if we don't need to reuse a txn?

fixed

**Comment by soundOfDestiny** [comment]
> IIUC, currently a new compaction group can only be created during table/mv creation (we can support group split via RPC in the future but just ignore that for now), which means compaction group creation always comes with a barrier. In this case, if we can make the following three operations atomic, group creation will never be reverted:
> 
> 1. `CompactionGroupManager::register_table_ids`
> 2. `HummockManager::sync_group`
> 3. `HummockManager::commit_epoch`
> 
> Given that CN can write table state without knowing the exact compaction group id, making the above operations atomic becomes possible. Here is my idea:
> 
> * Put a TableId to (CompcationGroupId, TableOption) map as part of the `CreateStreamingJob` [command](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/command.rs#L72)
> * After [completing MV creation barrier collection](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/mod.rs#L829), pass the above map into `commit_epoch`.
> * Inside `commit_epoch`, do the following operations in sequence:
> 
> 1. init a etcd transaction
> 2. register table ids in compaction group manager and put it into the transaction.
> 3. create GroupConstruct version delta, apply to hummock version and put it into the transaction.
> 4. execute existing logics in commit_epoch
> 5. commit the ectd transaction
> 
> With this, we can safely avoid reverting group creation during MV creation and avoid storing a ref to HummockManager in stream manager.

What if this barrier is without checkpoint?
If so, meta will not do `commit_epoch()` :-(

**Comment by zwang28** [comment]
> * Inside `commit_epoch`, do the following operations in sequence:

Even `commit_epoch` succeeds, following ops like [post_collect](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/mod.rs#L839) or [finish_commands](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/mod.rs#L868) can fail, thus [call for a revert](https://github.com/risingwavelabs/risingwave/blob/9be63c5a463de7716b82341edac65a656a8b2af1/src/meta/src/stream/stream_manager.rs#L673).

**Comment by hzxa21** [comment]
> > IIUC, currently a new compaction group can only be created during table/mv creation (we can support group split via RPC in the future but just ignore that for now), which means compaction group creation always comes with a barrier. In this case, if we can make the following three operations atomic, group creation will never be reverted:
> > 
> > 1. `CompactionGroupManager::register_table_ids`
> > 2. `HummockManager::sync_group`
> > 3. `HummockManager::commit_epoch`
> > 
> > Given that CN can write table state without knowing the exact compaction group id, making the above operations atomic becomes possible. Here is my idea:
> > 
> > * Put a TableId to (CompcationGroupId, TableOption) map as part of the `CreateStreamingJob` [command](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/command.rs#L72)
> > * After [completing MV creation barrier collection](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/mod.rs#L829), pass the above map into `commit_epoch`.
> > * Inside `commit_epoch`, do the following operations in sequence:
> > 
> > 
> > 1. init a etcd transaction
> > 2. register table ids in compaction group manager and put it into the transaction.
> > 3. create GroupConstruct version delta, apply to hummock version and put it into the transaction.
> > 4. execute existing logics in commit_epoch
> > 5. commit the ectd transaction
> > 
> > With this, we can safely avoid reverting group creation during MV creation and avoid storing a ref to HummockManager in stream manager.
> 
> What if this barrier is without checkpoint? If so, meta will not do `commit_epoch()` :-(

IIUC, completion of MV creation must be bound to a checkpoint epoch so commit_epoch will be called for sure.

**Comment by hzxa21** [comment]
> > * Inside `commit_epoch`, do the following operations in sequence:
> 
> Even `commit_epoch` succeeds, following ops like [post_collect](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/mod.rs#L839) or [finish_commands](https://github.com/risingwavelabs/risingwave/blob/572b53ec86d0949bac75745ce1954e2f2685f806/src/meta/src/barrier/mod.rs#L868) can fail, thus [call for a revert](https://github.com/risingwavelabs/risingwave/blob/9be63c5a463de7716b82341edac65a656a8b2af1/src/meta/src/stream/stream_manager.rs#L673).

You are right... I missed the fact that post operation of `commit_epoch` can fail and compaction group creation should be reverted. The root cause is under the current meta implementation, we cannot perform a single transaction containing all types of metadata updates (catalog, hummock version, compaction grpup, framgment, ...) and thus we use something similar to 2PC. In this case, revertion is unavoidable...

Not related to this PR: a better metadata management can be have a single MetadataManager management each sub-manager and we can commit metadata change belong to an epoch in one shot.

**Review by hzxa21 — APPROVED** [review]

**Pull request merged**